src/composed/message/parser.rs

```diff
@@ -39,21 +42,33 @@ fn next<I: Iterator<Item = Packet>>(packets: &mut Peekable<I>) -> Option<Result<
                         let mut edata = Vec::new();
 
                         // while ESK take em
-                        while packets.peek().map(|p| {
-                            p.tag() == Tag::PublicKeyEncryptedSessionKey
-                                || p.tag() == Tag::SymKeyEncryptedSessionKey
+                        while packets.peek().map(|p| match p.as_ref() {
+                            Ok(p) => {
+                                p.tag() == Tag::PublicKeyEncryptedSessionKey
+                                    || p.tag() == Tag::SymKeyEncryptedSessionKey
+                            }
+                            Err(_) => false,
                         }) == Some(true)
                         {
-                            esk.push(packets.next().expect("peeked").try_into().expect("peeked"));
+                            match packets.next().expect("peeked") {
+                                Ok(packet) => esk.push(packet.try_into().expect("peeked")),
+                                Err(e) => return Some(Err(e)),
+                            }
                         }
 
-                        // while edata take em
-                        while packets.peek().map(|p| {
-                            p.tag() == Tag::SymEncryptedData
-                                || p.tag() == Tag::SymEncryptedProtectedData
+                        // while edata take em (FIXME: the message grammar only allows one "Encrypted Data" packet)
+                        while packets.peek().map(|p| match p.as_ref() {
+                            Ok(p) => {
+                                p.tag() == Tag::SymEncryptedData
+                                    || p.tag() == Tag::SymEncryptedProtectedData
+                            }
+                            Err(_) => false,
```

Comment:
maybe write a helper function 

```rust
fn check_packet(f: F) -> bool 
where F: Fn(&Packet) -> bool {
  match packet {
    Ok(p) => f(p),
    Err(_) => false,
  }
}
```

Reply:
I think we're half re-inventing https://doc.rust-lang.org/std/result/enum.Result.html#method.is_ok_and (from Rust 1.70.0), there :sweat_smile: 

0ef10613701b3ae049c2b0c10a011b9a6c370433 is a bit less verbose, but not entirely satisfying. I keep wondering if I'm missing some additional simplification that can be done with `std`.

Reply:
sure

Reply:
are you voting to bump rust to 1.70 right away, or to change the code later?

Reply:
lets do it now